File: Civil_Engineering/Misc/FEM_Python/LinearTransient/AssembleStiffnessMatrix.py
#Assemble global stiffness matrix using element contributions
import numpy as np
from Local_Matrices import getStiffnessmatrix,getMassMatrix,getInternalForces
from getElementCoordinates import getElementCoordinates,getElementDisplacements,getElementCoordinatesIndex


def assemble(NodalCoord,Connectivity,D,D_prime,thickness,epsilon,sigma,el_type='Q4',P_type='vector',Prob='ES'):
    num_node = 4
    #now lets get the degrees of freedom (i.e. num_dof)
    if (P_type=='vector'):
        num_dof = num_node * 2  # elemental dof
        num_dof_total = 2 * len(NodalCoord)
        A = np.zeros((len(Connectivity),num_dof)) #create assembly matrix which is num_elements x num_dof
        for e in range(len(Connectivity)): #for each element
            for i in range(num_node): #all the columns of a row
                for j in range(2): #num of dof per node
                    if j==0:
                        A[e,(i)*2+j] = 2*Connectivity[e][i]-1 #assign x first based of nodal id of certain element hence
                    # con_mat[e,i]
                    else:
                        A[e,(i)*2+j] = 2*Connectivity[e][i] #assign y from nodal ID

    #print(A)
    K = np.zeros((num_dof_total, num_dof_total))  # K matrix will be num_dof x num_dof
    Fsig = np.zeros((num_dof_total,1))
    M = np.zeros((num_dof_total,num_dof_total)) # Mass matrix
    #Fext = np.zeros((num_dof_total,1))
    for e in range(len(Connectivity)):
        Coord_mat_el = getElementCoordinates(e,NodalCoord,Connectivity)
        Coord_mat_ind = getElementCoordinatesIndex(e,Connectivity)
        epsilon_el = epsilon[:, 4*e:4*e+4]
        sigma_el = sigma[:, 4*e:4*e+4]
        K_e = getStiffnessmatrix(Coord_mat_el,D,thickness,epsilon_el)
        Fsig_e = getInternalForces(Coord_mat_el,sigma_el)
        M_e = getMassMatrix(Coord_mat_el)
        for i in range(num_dof):
            dof_1 = int(A[e][i])-1 # get degrees of freedom
            Fsig[dof_1,0] += Fsig_e[i,0]
            #Fext[dof_1,0] += Fext_e[i,0]
            for j in range(num_dof):
                dof_2 = int(A[e][j])-1
                K[dof_1, dof_2] += K_e[i,j]
                M[dof_1,dof_2] += M_e[i,j]
            # end j loop
        #end i loop
    #end elemental loop
    return K,A,Fsig,M

File: Civil_Engineering/530_FEM/HW8/Matlab_routines/assembleStiffnessMatrix.m
function K = assembleStiffnessMatrix(NodalCoord, Connectivity, D, thickness)
%assembleStiffnessMatrix: This function assembles the global stiffness matrix using
%element contributions
% we will be using the getStiffnessMatrixQ4_PS function from the previous
% homework
    num_node = 4;
    num_dof = num_node*2; %per element
    num_dof_total = 12;
    A = zeros(length(Connectivity),num_dof); %create assembly matrix which is num_elements x num_dof
    
    for e = 1:2 %each element
         for i =1:num_node %all the columns of a row
             for j = 1:2 %num of dof per node
                % disp(j)
                  if (j==1)
                      A(e,(i-1)*2+j) = 2*Connectivity(e,i)-1 ; %x Dof
                    %disp(j)
                  else
                        A(e,(i-1)*2+j) = 2*Connectivity(e,i); %y Dof
                  end
             end %end j
         end %end i
         %disp(e)
    end%end e
%now to the actual assembly process
K = zeros(num_dof_total,num_dof_total);  % K matrix will be num_dof x num_dof
    %Loop over each element
    for e = 1:2
        Coords_el = getElementCoordinates(e, NodalCoord, Connectivity);
        Ke = getStiffnessMatrixQ4_PS(Coords_el, D, thickness);
        %print(K_e)

        %print(Coord_mat_el.T) %check that proper values are being returned (they are)

        for i= 1:num_dof
            dof_1 = A(e,i);  % get degrees of freedom
            for j = 1:num_dof
                dof_2 = A(e,j);
                K(dof_1, dof_2) = K(dof_1,dof_2) +  Ke(i,j);
            % end j loop
            end
        %end i loop
        end
%end elemental loop
    end
end
    



